sopel/tools/memories.py

```diff
@@ -161,3 +230,57 @@ def __contains__(self, key):
 
     def __setitem__(self, key: Optional[str], value):
         super().__setitem__(self._make_key(key), value)
+
+    def setdefault(self, key: str, default=None):
+        return super().setdefault(self._make_key(key), default)
+
+    def __delitem__(self, key: str):
+        super().__delitem__(self._make_key(key))
+
+    def copy(self):
+        return type(self)(self, identifier_factory=self.make_identifier)
+
+    def get(self, key: str, default=_NO_DEFAULT):
+        if default is _NO_DEFAULT:
+            return super().get(self._make_key(key))
+        return super().get(self._make_key(key), default)
+
+    def pop(self, key: str, default=_NO_DEFAULT):
+        if default is _NO_DEFAULT:
+            return super().pop(self._make_key(key))
+        return super().pop(self._make_key(key), default)
+
+    def update(self, maybe_mapping=tuple()):
+        super().update(self._convert_keys(maybe_mapping))
+
+    def __or__(self, other):
+        if not isinstance(other, dict):
+            return NotImplemented
+
+        # self on the left, so other's keys overwrite
+        new = self.copy()
+        new.update(other)
+        return new
+
+    def __ror__(self, other):
+        if not isinstance(other, dict):
+            return NotImplemented
+
+        # self on the right, so keep only new keys from other
+        new = self.copy()
+        new.update((k, v) for k, v in other.items() if k not in self)
```

Comment:
Am I mistaken, or can this be done more symmetrically by copying `other` first? Keys from `self` should override any shared keys, and everything else is in the disjoint and belongs in the union.

```suggestion
        # self on the right, so keep only new keys from other
        new = other.copy()
        new.update(self)
```

Reply:
[Deja vu](https://github.com/sopel-irc/sopel/pull/2525#discussion_r1369588596)? Previous thread tl;dr: `other` might be a normal `dict`, and a copy of it wouldn't be a `SopelIdentifierMemory` if that's the case. That's how we got to the `new = type(self)(other)` construct that @Exirel proposed replacing with a `new = self.copy()` + the generator-expression-based `new.update()` call.

Reply:
d'oh, right